proposals/astro-request.md

```diff
@@ -0,0 +1,110 @@
+- Start Date: 2022-03-21
+- Reference Issues: https://github.com/withastro/rfcs/discussions/151
+- Implementation PR: <!-- leave empty -->
+
+# Summary
+
+- Change `Astro.request` to become a [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) object.
+- Move `Astro.request.params` to `Astro.params`.
+- Move `Astro.request.canonicalURL` to `Astro.canonicalURL`.
+
+# Example
+
+```astro
+---
+const cookie = Astro.request.headers.get('cookie');
+const loggedIn = parse(cookie).loggedIn;
+---
+<html>
+<head>
+  <title>My Blog</title>
+  <link rel="canonical" href={Astro.canonicalURL}>
+</head>
+<body>
+  Hello, you are {!loggedIn && 'not'} logged in.
+</body>
+</html>
+```
+
+# Motivation
+
+In server-side rendering contexts you will want access to the HTTP request headers to do things such as check cookies.
+
+[Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) is a standard API to represent an HTTP request that is used by:
+
+- Service workers
+- Cloudflare Workers
+- Deno
+
+As well as other meta-frameworks:
+
+- SvelteKit
+- Remix
+
+The Request object allows access to headers through the `request.headers` Map-like object. Switching `Astro.request` to a Request will provide a familiar interface for this use-case.
+
+# Detailed design
+
+__Astro.request__ is currently an object with this interface:
+
+```typescript
+interface AstroRequest {
+  /** get the current page URL */
+  url: URL;
+
+  /** get the current canonical URL */
+  canonicalURL: URL;
+
+  /** get page params (dynamic pages only) */
+  params: Params;
+}
+```
+
+This change will move `canonicalURL` and `params` up to the `Astro` object and make `request` a Request.
+
+```typescript
+// Partial
+interface Astro {
+    /** get the current canonical URL */
+  canonicalURL: URL;
+
+  /** get page params (dynamic pages only) */
+  params: Params;
```

Comment:
You mentioned below that `Astro.request` will have all query/search params stripped in SSG mode, but I can imagine I'd be confused if `Astro.params` is still populated and the two don't match. 

Instead of stripping these out in SSG, should we just clean them up to match only the params supported by the file-based route? We already implement this internally, so it wouldn't be any more work (I don't think, at least).

Or, if we want to go full purist/breakage, we could remove Astro.params entirely in favor of `Astro.request.searchParams` in the SSG use-case. That feels a bit more less user-friendly though, since `searchParams` isn't the most well-known API.

Reply:
Astro.params is the file based params, it's not related to search/query parameters. I can't imagine being confused by this, as this is currently how it works already.